#' Script to create three subplots for figure 4.
#' 
#' 1. Plot the distances between accessions with the minor allele at each SNP,
#' along with the distribution of distances between the same number of pairs of
#' randomly chosen SNPs.
#' 2. Plot the distances between accessions within each admixture group,
#' along with the distribution of distances between the same number of pairs of
#' randomly chosen accessions.
#' 3. Minor allele frequencies within each admixture group

# Functions for sampling from distance matrices
source("02_library/matrix_subsamples.R")
# Import data
source('03_data_preparation/1001genomes_data.R')

# Add a column for genotype at the top SNP
g1001 <- g1001 %>% 
  left_join(
    # read_csv("01_data/top_snp.csv", col_types = "ci"),
    read_csv("05_figures/fig2/relative_risk/output/snp_matrix.csv", col_types = "ciiiiiiiiiiiii"),
    by = 'code'
  )

# Summary tables generated by `04_analyses/07_geographic_distribution/distance_by_MAC.py`
# Import summary table for distances between susceptible alleles at the 13 loci
obs_loci  <- read_csv("04_analyses/07_geographic_distribution/output/distances_between_observed_loci.csv")
# Import summary of minor allele counts, and mean distance between accessions with the minor allele
# for 10,000 randomly chosen alleles.
rand_loci <- read_csv("04_analyses/07_geographic_distribution/output/distances_between_random_loci.csv")

obs_loci <- obs_loci %>% 
  mutate(
    label  = ifelse(mean_distance > 3000, paste0("Chr. ", chr, ":", pos), "")
  )

mac_by_distance <- rand_loci %>%
  ggplot(aes( x = mac, y = mean_distance)) +
  geom_point(color = 'gray', size = 0.1) +
  geom_point(data = obs_loci, color = 'red') + 
  geom_text(data = obs_loci, aes( x = mac, y = mean_distance, label = label),
            hjust=0.5,vjust=-1,
            # angle = 45
            size = 2
  ) +
  labs(
    x = "Minor allele count",
    y = "Mean pairwise distance (km)"
  )

# # Summary tables generated by `04_analyses/07_geographic_distribution/test_spatial_dispersion.R`
# # Import summary table for distances between susceptible alleles at the 13 loci
# snp_dists <- read_csv(
#   "04_analyses/07_geographic_distribution/output/snps_distances.csv",
#   col_types = cols()
# )
# # Import summary table for distances between accessions from each admixture group
# am_dists <- read_csv(
#   "04_analyses/07_geographic_distribution/output/admixture_group_distances.csv",
#   col_types = cols()
# ) %>% 
#   mutate(am_group = recode(
#     am_group, 
#     western_europe = "W. Europe",
#     admixed = "Admixed",
#     central_europe = "C. Europe",
#     germany = "Germany",
#     asia = "Asia",
#     south_sweden = "S. Sweden",
#     north_sweden = "N. Sweden",
#     relict = "Relict",
#     spain = "Spain",
#     italy_balkan_caucasus = "It./Balk./Cauc."
#   )
#   )
# 
# # Plot distances between alleles
# plot_snp_dists <- snp_dists %>% 
#   ggplot(aes(x = snp, y = obs)) +
#   geom_point() +
#   geom_errorbar(aes(ymin = lower, ymax = upper, width=0.1)) +
#   labs(
#     y = "Distance (km)",
#     x = ""
#   ) +
#   theme_bw() +
#   theme(
#     axis.text.x = element_text(angle = 45, hjust = 1)
#   )
# 
# # Plot distances between accessions in each admixture group
# plot_am_dists <- am_dists %>% 
#   ggplot(aes(x = am_group, y = obs)) +
#   geom_point() +
#   geom_errorbar(aes(ymin = lower, ymax = upper, width=0.1)) +
#   labs(
#     y = "Distance (km)",
#     x = "Admixture group"
#   ) +
#   theme_bw() +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Plot minor allele frequency of the major SNP within each admixture group.
plot_maf <- g1001 %>% 
  group_by(admixture_group) %>% 
  summarise(
    af = mean(chr2_5927469),
    st_err = binomial_st_err(chr2_5927469),
    .groups = 'drop'
  ) %>% 
  mutate(admixture_group = recode(
    admixture_group, 
    western_europe = "W. Europe",
    admixed = "Admixed",
    central_europe = "C. Europe",
    germany = "Germany",
    asia = "Asia",
    south_sweden = "S. Sweden",
    north_sweden = "N. Sweden",
    relict = "Relict",
    spain = "Spain",
    italy_balkan_caucasus = "It./Balk./Cauc."
  )) %>% 
  ggplot(aes(x = admixture_group, y = af)) +
  geom_point() +
  geom_errorbar(aes(ymin=af-st_err, ymax = af+st_err), width = 0.1) +
  geom_hline(yintercept = mean(g1001$chr2_5927469), colour = 'gray70', linetype='dashed') +
  labs(
    x = "Admixture group",
    y = "Allele frequency"
  ) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


