---
title: "Supplementary figures"
# author: "Tom Ellis"
date: ""
output:
  bookdown::pdf_book:
    toc: no
    keep_tex: no
    number_sections: false
    latex_engine: lualatex
  bookdown::word_document2: null
---
\renewcommand{\thefigure}{S\arabic{figure}}
\setcounter{figure}{0}
\renewcommand{\thetable}{S\arabic{table}}
\setcounter{table}{0}
\renewcommand{\theequation}{S\arabic{equation}}
\setcounter{equation}{0}

This document contains supplementary figures to accompany the manuscript "A major virus-resistance association in *Arabidopsis thaliana* is consistent with negative frequency-dependent selection".

```{r supp-setup, echo=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  cache=TRUE, cache.lazy=TRUE,
  fig.width=16.9/2.54, fig.height = 10/2.54, fig.align='center',
  warning = F, message =F,
  dev = "png"
)

suppressPackageStartupMessages(library(tidyverse))
source("02_library/manhattan_plot.R")
```

(ref:snp-heritability) SNP heritability in response to each virus.

```{r snp-heritability, fig.cap="(ref:snp-heritability)"}
# Import pseudoheritabilities and var. expained by the top SNP
# generated by pseudoheritabiliy.Rmd
h2 <- read_csv("04_analyses/01_snp_heritability/snp_heritability.csv", col_types = 'ccdd')

pd <- position_dodge(0.2)

h2 %>%
  mutate(
    Virus = ifelse(grepl("*Anc*", trait), "Ancestral", "Evolved"),
    trait = str_extract(trait, "^[^_]+"),
    trait = ifelse(trait == "SYM", "Symptoms", trait)
  ) %>%
  ggplot(aes(x = trait, y = Estimate, group=paste(Virus, component), colour = Virus)) +
  geom_point(position = pd) +
  geom_errorbar(aes(
    ymin = Estimate - SE,
    ymax =  Estimate + SE
  ),
  position = pd, width = 0.1) +
  labs(
    x = "",
    y = "SNP heritability"
  ) +
  theme(
    axis.title.x = element_blank(),
    legend.position = "top"
    ) +
  scale_x_discrete(
    labels =
      c("AUDPS", "Infectivity", "Necrosis", "Severity of\nsymptoms")
  ) +
  theme_bw()


```

(ref:replicate-expt) Manhattan plots for the association between SNPs and necrosis in the replicate dataset of 118 accessions.

```{r replicate-expt, fig.cap="(ref:replicate-expt)", fig.width=16.9/2.54, fig.height=10/2.54}
path <- "04_analyses/04_replicate_dataset/output/"
read_csv(
    paste(path, "binary_necrosis_anc_binary_necrosis_evo_0.03_MTMM_G.csv", sep = ""),
    col_types = 'ciddid'
  ) %>% 
  manhattan_plot()

```


(ref:g-manh) Manhattan plots for associations with common disease responses to both viruses.

```{r g-manh, fig.cap="(ref:g-manh)", fig.height=20/2.54}
# Import files for G results
limix_path <- "04_analyses/02_multitrait_GWA/output/"
limix_files <- list.files(limix_path, "^[ASI].*G.csv")
limix_files <-paste(limix_path, limix_files, sep ="")

limix_G <- lapply(limix_files, read_csv, col_types = 'ciddid')
names(limix_G) <- c("audps", "inf", "sym")

ggpubr::ggarrange(
  manhattan_plot(limix_G$audps) + ggtitle("AUDPS"),
  manhattan_plot(limix_G$inf)   + ggtitle("Infectivity"),
  manhattan_plot(limix_G$sym)   + ggtitle("Severity of symptoms"),
  nrow=3, ncol= 1, labels = "AUTO"
)

```

(ref:gxe-manh) Manhattan plots for associations with disease responses specific to individual viral isolates.

```{r gxe-manh, fig.cap="(ref:gxe-manh)", fig.height=20/2.54}

# Import files for GxE results
limix_path <- "04_analyses/02_multitrait_GWA/output/"
limix_files <- list.files(limix_path, "^[ASI].*GxE.csv")
limix_files <-paste(limix_path, limix_files, sep ="")

limix_GxE <- lapply(limix_files, read_csv, col_types = 'ciddid')
names(limix_GxE) <- c("audps", "inf", "necrosis", "sym")


ggpubr::ggarrange(
  manhattan_plot(limix_GxE$audps) + ggtitle("AUDPS"),
  manhattan_plot(limix_GxE$inf) + ggtitle("Infectivity"),
  manhattan_plot(limix_GxE$sym) + ggtitle("Severity of symptoms"),
  nrow=3, ncol= 1, labels = "AUTO"
)

```

(ref:ld-matrix) Linkage disequilibrium between SNPs associated with necrosis.

```{r, fig.cap="(ref:ld-matrix)", fig.width=16.9/2.54, fig.height=20/2.54}
source("03_data_preparation/1001genomes_data.R")
g1001 <- g1001 %>% 
  left_join(
    read_csv("05_figures/fig2/relative_risk/output/snp_matrix.csv", col_types = 'ciiiiiiiiiiiii'),
    by='code'
  ) %>% 
  mutate(
    chr4_273465 = 1 - chr4_273465, # So the minor allele is the susceptible allele
  )
# Pull the vector of SNP names from the column headers
snp_names <- grep("chr._", colnames(g1001), value = TRUE)

# Get upper triangle of the correlation matrix
get_upper_tri <- function(cormat){
  cormat[lower.tri(cormat, diag = TRUE)]<- NA
  return(cormat)
}

cor(g1001[, snp_names], method = 's') %>%
  get_upper_tri() %>%
  reshape2::melt(na.rm = TRUE) %>%
  ggplot(aes(Var1, Var2, fill=value)) +
  geom_tile(colour="white") +
  geom_text(aes(Var1, Var2, label = round(value,2)), color = "black", size = 3) +
  scale_fill_gradient2(
    low = "blue", high = "red", mid = "white",
    midpoint = 0, limit = c(-1,1), space = "Lab",
    name="Spearman\nCorrelation") +
  theme_minimal()+
  coord_fixed() +
  theme(
    legend.position="none",
    axis.title = element_blank(),
    axis.text.x = element_text( angle = 45, vjust = 1, hjust = 1)
  ) +
  labs(
    x = "",
    y = ""
  )

```


(ref:kruskal-wallis) Manhattan plots for the association between SNPs and necrosis using a Kruskall-Wallis test at each locus, with no correction for population structure.

```{r kruskal-wallis, fig.cap="(ref:kruskal-wallis)", eval=F}

path <- "005_results/017_kruskall-wallis_gwas/output/"

# kw_anc <- read_csv_chunked(
#   paste(path, "ancestral.csv", sep = ""),
#   callback = DataFrameCallback$new(function(x, pos) subset(x, maf > 0.1)),
#   col_types = 'ciddid',
#   progress = F) %>% 
#   manhattan_plot() +
#     ggtitle("TuMV-Anc")


kw_evo <- read_csv_chunked(
  paste(path, "evolved.csv", sep = ""),
  callback = DataFrameCallback$new(function(x, pos) subset(x, maf > 0.1)),
  col_types = 'ciddid',
  progress = F) %>%
  manhattan_plot() +
    ggtitle("TuMV-Evo")


# ggpubr::ggarrange(
#   kw_anc, kw_evo,
#   ncol= 2, labels = "AUTO"
# )




```

