---
title: "Supplementary figures"
# author: "Tom Ellis"
date: ""
output:
  bookdown::pdf_book:
    toc: no
    keep_tex: no
    number_sections: false
  bookdown::word_document2: null
---
\renewcommand{\thefigure}{S\arabic{figure}}
\setcounter{figure}{0}
\renewcommand{\thetable}{S\arabic{table}}
\setcounter{table}{0}
\renewcommand{\theequation}{S\arabic{equation}}
\setcounter{equation}{0}

This document contains supplementary figures to accompany the manuscript "Genetic basis of *Arabidopsis thaliana* responses to infection by naïve and adapted isolates of turnip mosaic virus" by Anamarija Butković, Thomas James Ellis, Rubén González, Benjamin Jaegle, Magnus Nordborg, and Santiago F. Elena.

```{r supp-setup, echo=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  cache=TRUE, cache.lazy=TRUE,
  warning = FALSE, message =FALSE
)
# This gets the PNG device to work on RStudio server
# options(bitmapType='cairo')

suppressPackageStartupMessages(library(tidyverse))
library(ggpubr)
source("02_library/manhattan_plot.R")
```

<!-- (ref:necrosis-snps) Loci associated with necrosis, showing whether the association was with the ‘common’ response to both viruses, or with a virus-specific response. -->

<!-- ```{r necrosis-snps} -->
<!-- read_csv("05_figures/supplementary_figures/supplementary_table_1.csv") %>%  -->
<!--   knitr::kable(caption = "(ref:necrosis-snps)") -->
<!-- ``` -->

<!-- (ref:snp-heritability) SNP heritability in response to each virus. -->

<!-- ```{r snp-heritability, fig.cap="(ref:snp-heritability)"} -->
<!-- # Import pseudoheritabilities and var. expained by the top SNP -->
<!-- # generated by pseudoheritabiliy.Rmd -->
<!-- h2 <- read_csv("04_analyses/01_snp_heritability/snp_heritability.csv", col_types = 'ccdd') -->

<!-- pd <- position_dodge(0.2) -->

<!-- h2 %>% -->
<!--   mutate( -->
<!--     Virus = ifelse(grepl("*Anc*", trait), "Ancestral", "Evolved"), -->
<!--     trait = str_extract(trait, "^[^_]+"), -->
<!--     trait = ifelse(trait == "SYM", "Symptoms", trait) -->
<!--   ) %>% -->
<!--   ggplot(aes(x = trait, y = Estimate, group=paste(Virus, component), colour = Virus)) + -->
<!--   geom_point(position = pd) + -->
<!--   geom_errorbar(aes( -->
<!--     ymin = Estimate - SE, -->
<!--     ymax =  Estimate + SE -->
<!--   ), -->
<!--   position = pd, width = 0.1) + -->
<!--   labs( -->
<!--     x = "", -->
<!--     y = "SNP heritability" -->
<!--   ) + -->
<!--   theme( -->
<!--     axis.title.x = element_blank(), -->
<!--     legend.position = "top" -->
<!--     ) + -->
<!--   scale_x_discrete( -->
<!--     labels = -->
<!--       c("AUDPS", "Infectivity", "Necrosis", "Severity of\nsymptoms") -->
<!-- ) + -->
<!--   theme_bw() -->


<!-- ``` -->

(ref:replicate-expt) Manhattan plots for the association between SNPs and necrosis in the replicate dataset of 118 accessions.

```{r replicate-expt, fig.cap="(ref:replicate-expt)", fig.width=16.9/2.54, fig.height=10/2.54}
path <- "04_analyses/04_replicate_dataset/output/"
read_csv(
    paste(path, "binary_necrosis_anc_binary_necrosis_evo_0.03_MTMM_G.csv", sep = ""),
    col_types = 'ciddid'
  ) %>% 
  manhattan_plot()

```

(ref:conditional-GWA) Manhattan plots for associations with severity and symptoms and necrosis in a model conditioned on the most strongly associated SNP (Chr2:5927469).

```{r conditional-GWA, fig.cap="(ref:conditional-GWA)", fig.width=16.9/2.54, fig.height=15/2.54}
gwa_filenames <- c(
  "04_analyses/03_condition_on_top_snp/output/SYM_Anc_SYM_Evo_0.03_MTMM_G.csv",
  "04_analyses/03_condition_on_top_snp/output/SYM_Anc_SYM_Evo_0.03_MTMM_GxE.csv",
  "04_analyses/03_condition_on_top_snp/output/Necrosis_Evo_Necrosis_Anc_0.03_MTMM_G.csv",
  "04_analyses/03_condition_on_top_snp/output/Necrosis_Evo_Necrosis_Anc_0.03_MTMM_GxE.csv"
)
gwa_results <- lapply(gwa_filenames, read_csv, col_types = 'ciddid')
names(gwa_results) <- c("symptoms_G", "symptoms_GxE", "necrosis_G", "necrosis_GxE")

# For each GWA file add x-axis positions and filter only the significant results
signif_snps <- lapply(gwa_results, function(x){
  x %>% 
    add_base_pair_positions(.) %>% 
    mutate(log10p = -log10(pvalue)) %>% 
    filter( log10p > -log10(0.05 / nrow(.)) ) 
})

# Base Manhattan plots
manh <- lapply(gwa_results, manhattan_plot,
               fraction_to_keep = 0.05, signif_cutoff = 0.05 , chr_colours = c('gray80', 'gray60')
)
# Label significant SNPs in red
manh_plus_signif <- list(
  manh$symptoms_G +
    geom_point(
      data = signif_snps$symptoms_G, inherit.aes = TRUE, colour = 'red'
    )+
    ggtitle("Severity of symptoms", "Common effects"),
  manh$necrosis_G +
    geom_point(
      data = signif_snps$necrosis_G, inherit.aes = TRUE, colour = 'red'
    ) +
    ggtitle("Necrosis", "Common effects"),
  manh$symptoms_GxE +
    geom_point(
      data = signif_snps$symptoms_GxE, inherit.aes = TRUE, colour = 'red'
    )+
    ggtitle("Severity of symptoms", "Virus-specific effects"),
  manh$necrosis_GxE +
    geom_point(
      data = signif_snps$necrosis_GxE, inherit.aes = TRUE, colour = 'red'
    ) +
    ggtitle("Necrosis", "Virus-specific effects")
)

ggarrange(
  plotlist = manh_plus_signif,
  ncol=2, nrow = 2, labels = "AUTO"
)

```

(ref:ld-matrix) Linkage disequilibrium between SNPs associated with necrosis.

```{r ld-matrix, fig.cap="(ref:ld-matrix)", fig.width=16.9/2.54, fig.height=20/2.54}
source("03_data_preparation/1001genomes_data.R")
g1001 <- g1001 %>% 
  left_join(
    read_csv("05_figures/fig2/relative_risk/output/snp_matrix.csv", col_types = 'ciiiiiiiiiiiii'),
    by='code'
  ) %>% 
  mutate(
    chr4_273465 = 1 - chr4_273465, # So the minor allele is the susceptible allele
  )
# Pull the vector of SNP names from the column headers
snp_names <- grep("chr._", colnames(g1001), value = TRUE)

# Get upper triangle of the correlation matrix
get_upper_tri <- function(cormat){
  cormat[lower.tri(cormat, diag = TRUE)]<- NA
  return(cormat)
}

cor(g1001[, snp_names], method = 's') %>%
  get_upper_tri() %>%
  reshape2::melt(na.rm = TRUE) %>%
  ggplot(aes(Var1, Var2, fill=value)) +
  geom_tile(colour="white") +
  geom_text(aes(Var1, Var2, label = round(value,2)), color = "black", size = 3) +
  scale_fill_gradient2(
    low = "blue", high = "red", mid = "white",
    midpoint = 0, limit = c(-1,1), space = "Lab",
    name="Spearman\nCorrelation") +
  theme_minimal()+
  coord_fixed() +
  theme(
    legend.position="none",
    axis.title = element_blank(),
    axis.text.x = element_text( angle = 45, vjust = 1, hjust = 1)
  ) +
  labs(
    x = "",
    y = ""
  )

```

(ref:audps-infectivity-gwa) Manhattan plots for common and virus-specific associations with AUDPS and infectivity.

```{r audps-infectivity-gwa, fig.cap="(ref:audps-infectivity-gwa)", fig.width=16.9/2.54, fig.height=20/2.54}

gwa_filenames <- c(
  "04_analyses/02_multitrait_GWA/output/AUDPS_Anc_21_AUDPS_Evo_21_0.03_MTMM_G.csv",
  "04_analyses/02_multitrait_GWA/output/Infectivity_Anc_21_Infectivity_Evo_21_0.03_MTMM_G.csv",
  "04_analyses/02_multitrait_GWA/output/AUDPS_Anc_21_AUDPS_Evo_21_0.03_MTMM_GxE.csv",
  "04_analyses/02_multitrait_GWA/output/Infectivity_Anc_21_Infectivity_Evo_21_0.03_MTMM_GxE.csv"
)
gwa_results <- lapply(gwa_filenames, read_csv, col_types = 'ciddid')
names(gwa_results) <- c("audps_G", "infectivity_G", "audps_GxE", "infectivity_GxE")

# Base Manhattan plots
manh <- lapply(gwa_results, manhattan_plot,
               fraction_to_keep = 0.05, signif_cutoff = 0.05 , chr_colours = c('gray80', 'gray60')
)
# Label significant SNPs in red
audps_inf_manh <- list(
  manh$audps_G +
    ggtitle("AUDPS", "Common effects"),
  manh$infectivity_G +
    ggtitle("Infectivity", "Common effects"),
  manh$audps_GxE +
    ggtitle("AUDPS", "Virus-specific effects"),
  manh$infectivity_GxE +
    ggtitle("Infectivity", "Virus-specific effects")
)

ggarrange(
  plotlist = audps_inf_manh,
  ncol=2, nrow = 2, labels = "AUTO"
)

```

(ref:kruskal-wallis) Manhattan plots for the association between SNPs and necrosis using a Kruskall-Wallis test at each locus, with no correction for population structure.

```{r kruskal-wallis, fig.cap="(ref:kruskal-wallis)", eval=F}

path <- "005_results/017_kruskall-wallis_gwas/output/"

# kw_anc <- read_csv_chunked(
#   paste(path, "ancestral.csv", sep = ""),
#   callback = DataFrameCallback$new(function(x, pos) subset(x, maf > 0.1)),
#   col_types = 'ciddid',
#   progress = F) %>% 
#   manhattan_plot() +
#     ggtitle("TuMV-Anc")


kw_evo <- read_csv_chunked(
  paste(path, "evolved.csv", sep = ""),
  callback = DataFrameCallback$new(function(x, pos) subset(x, maf > 0.1)),
  col_types = 'ciddid',
  progress = F) %>%
  manhattan_plot() +
    ggtitle("TuMV-Evo")


# ggpubr::ggarrange(
#   kw_anc, kw_evo,
#   ncol= 2, labels = "AUTO"
# )




```

